### PR TITLE
Expr simplifier: simplification passes for matmul

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Warning: this PR contains https://github.com/csarofeen/pytorch/pull/2258 and https://github.com/csarofeen/pytorch/pull/2273. Please review this PR after I have merged these two PRs and rebased this PR.

This PR adds a few more passes that are capable of simplifying matmul indexing well. The newly added passes are: `cancelDivMod`, `distributeDivisibleDivMod`, and `distributeMul`. The most helpful pass for matmul is `distributeDivisibleDivMod`. It simplifies indices like:
```CUDA
(threadIdx.x + 16 * i1) % 8
```
into
```CUDA
threadIdx.x % 8
```
which helps removing data dependency on `i1` so that the index can be hoisted outside of the `i1` loop.

# Example matmul kernel code

Command:
```
$CUDA_VISIBLE_DEVICES=1 $PYTORCH_NVFUSER_DUMP="cuda_kernel,expr_simplify,ptxas_verbose" ./build/bin/nvfuser_bench --benchmark_filter=Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_NT_Legacy/2048/3456/4096/manual_time
```

Kernel diff compare (this PR + https://github.com/csarofeen/pytorch/pull/1900) vs https://github.com/csarofeen/pytorch/pull/1900 alone
https://www.diffchecker.com/pcxGCQkn

# Matmul perf benchmark

Compare (this PR + https://github.com/csarofeen/pytorch/pull/1900) vs https://github.com/csarofeen/pytorch/pull/1900 alone

Command:
```
$CUDA_VISIBLE_DEVICES=1 ./build/bin/nvfuser_bench --benchmark_filter=.*Matmul.*Legacy/2048/3456/4096.*
```

Before:
```C++
---------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                       Time             CPU   Iterations
---------------------------------------------------------------------------------------------------------------------------------
Nvfuser_Matmul_4warp3stage/no_quant_nvfuser_4warp_TT_Legacy/2048/3456/4096/manual_time       1209 us         1387 us          527
Nvfuser_Matmul_4warp3stage/no_quant_nvfuser_4warp_TN_Legacy/2048/3456/4096/manual_time       1204 us         1382 us          504
Nvfuser_Matmul_4warp3stage/no_quant_nvfuser_4warp_NT_Legacy/2048/3456/4096/manual_time       1126 us         1304 us          538
Nvfuser_Matmul_4warp4stage/no_quant_nvfuser_4warp_TT_Legacy/2048/3456/4096/manual_time       2018 us         2200 us          312
Nvfuser_Matmul_4warp4stage/no_quant_nvfuser_4warp_TN_Legacy/2048/3456/4096/manual_time       2082 us         2271 us          273
Nvfuser_Matmul_4warp4stage/no_quant_nvfuser_4warp_NT_Legacy/2048/3456/4096/manual_time       1957 us         2138 us          305
Nvfuser_Matmul_8warp3stage/no_quant_nvfuser_8warp_TT_Legacy/2048/3456/4096/manual_time       1297 us         1491 us          460
Nvfuser_Matmul_8warp3stage/no_quant_nvfuser_8warp_TN_Legacy/2048/3456/4096/manual_time       1358 us         1580 us          460
Nvfuser_Matmul_8warp3stage/no_quant_nvfuser_8warp_NT_Legacy/2048/3456/4096/manual_time       1314 us         1493 us          454
Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_TT_Legacy/2048/3456/4096/manual_time       1411 us         1590 us          423
Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_TN_Legacy/2048/3456/4096/manual_time       1407 us         1587 us          424
Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_NT_Legacy/2048/3456/4096/manual_time       1318 us         1497 us          453
EagerModeMatmul/no_quant_eagermode_TT_Legacy/2048/3456/4096/manual_time                       878 us          942 us          806
EagerModeMatmul/no_quant_eagermode_TN_Legacy/2048/3456/4096/manual_time                       904 us          971 us          747
EagerModeMatmul/no_quant_eagermode_NT_Legacy/2048/3456/4096/manual_time                       834 us          903 us          856

```

After:
```C++
---------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                       Time             CPU   Iterations
---------------------------------------------------------------------------------------------------------------------------------
Nvfuser_Matmul_4warp3stage/no_quant_nvfuser_4warp_TT_Legacy/2048/3456/4096/manual_time        929 us         1109 us          678
Nvfuser_Matmul_4warp3stage/no_quant_nvfuser_4warp_TN_Legacy/2048/3456/4096/manual_time        951 us         1138 us          655
Nvfuser_Matmul_4warp3stage/no_quant_nvfuser_4warp_NT_Legacy/2048/3456/4096/manual_time        893 us         1083 us          680
Nvfuser_Matmul_4warp4stage/no_quant_nvfuser_4warp_TT_Legacy/2048/3456/4096/manual_time       1066 us         1249 us          595
Nvfuser_Matmul_4warp4stage/no_quant_nvfuser_4warp_TN_Legacy/2048/3456/4096/manual_time       1177 us         1358 us          529
Nvfuser_Matmul_4warp4stage/no_quant_nvfuser_4warp_NT_Legacy/2048/3456/4096/manual_time       1032 us         1217 us          604
Nvfuser_Matmul_8warp3stage/no_quant_nvfuser_8warp_TT_Legacy/2048/3456/4096/manual_time        936 us         1117 us          669
Nvfuser_Matmul_8warp3stage/no_quant_nvfuser_8warp_TN_Legacy/2048/3456/4096/manual_time        911 us         1182 us          670
Nvfuser_Matmul_8warp3stage/no_quant_nvfuser_8warp_NT_Legacy/2048/3456/4096/manual_time        928 us         1109 us          670
Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_TT_Legacy/2048/3456/4096/manual_time        914 us         1117 us          671
Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_TN_Legacy/2048/3456/4096/manual_time        936 us         1115 us          668
Nvfuser_Matmul_8warp4stage/no_quant_nvfuser_8warp_NT_Legacy/2048/3456/4096/manual_time        926 us         1108 us          669
EagerModeMatmul/no_quant_eagermode_TT_Legacy/2048/3456/4096/manual_time                       876 us          957 us          811
EagerModeMatmul/no_quant_eagermode_TN_Legacy/2048/3456/4096/manual_time                       903 us          977 us          755
EagerModeMatmul/no_quant_eagermode_NT_Legacy/2048/3456/4096/manual_time                       829 us          905 us          840
```

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
@naoyam This is ready for review

File: third_party/nvfuser/test/test_gpu_expr_simplifier.cpp [hunk 1]
```diff
@@ -13,7 +13,52 @@ namespace jit {
...
+      simplified->toInlineString());
+}
+
+} // namespace
+
 TEST_F(NVFuserTest, FusionAssociativeAndCommutativeReordering_CUDA) {
+  // We need simplifications like a + b == a + b --> true to pass this test
+  // For now skip.
+  // TODO: fix this
+  return;
```
Comment (zasdfgbnm) [inline third_party/nvfuser/test/test_gpu_expr_simplifier.cpp:61]: Looks like I missed this failure last year. Will write a separate PR adding the missing feature. We can hold this PR until that is done.

**Comment by naoyam** [comment]
> This PR adds a few more passes that are capable of simplifying matmul indexing well. The newly added passes are: `cancelDivMod`, `distributeDivisibleDivMod`, and `distributeMul`. The most helpful pass for matmul is `distributeDivisibleDivMod`. It simplifies indices like:
> 
> ```cuda-c++
> (threadIdx.x + 16 * i1) % 8
> ```
> 
> into
> 
> ```cuda-c++
> threadIdx.x
> ```
> 

Why is this legal? Is threadIdx.x assumed to be less than 8?

File: third_party/nvfuser/csrc/expr_simplifier.cpp [hunk 2]
```diff
@@ -753,6 +753,17 @@ bool isFlattenedMul(Val* x) {
   return toFlattenedMul(x->definition()) != nullptr;
 }
 
+BinaryOp* toDivModOp(Expr* expr) {
+  if (auto bop = dynamic_cast<BinaryOp*>(expr)) {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/expr_simplifier.cpp]: nit: Most of the functions have this pattern of conditional branches where we could reduce indentation levels by negating the condition and exit.

**Comment by zasdfgbnm** [comment]
> > This PR adds a few more passes that are capable of simplifying matmul indexing well. The newly added passes are: `cancelDivMod`, `distributeDivisibleDivMod`, and `distributeMul`. The most helpful pass for matmul is `distributeDivisibleDivMod`. It simplifies indices like:
> > ```
> > (threadIdx.x + 16 * i1) % 8
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > into
> > ```
> > threadIdx.x
> > ```
> 
> Why is this legal? Is threadIdx.x assumed to be less than 8?

Oh, sorry. I meant to say `threadIdx.x % 8`...

**Review by naoyam — APPROVED** [review]
LGTM

File: third_party/nvfuser/csrc/expr_simplifier.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/expr_simplifier.cpp]: changed most of them

**Pull request merged**